sopel/modules/url.py

```diff
@@ -167,20 +148,22 @@ def title_auto(bot, trigger):
         if bot.memory['safety_cache'][trigger]['positives'] > 1:
             return
 
-    urls = find_urls(trigger, clean=True)
-    if len(urls) == 0:
-        return
-
-    results = process_urls(bot, trigger, urls)
-    bot.memory['last_seen_url'][trigger.sender] = urls[-1]
+    urls = web.search_urls(
+        trigger, exclusion_char=bot.config.url.exclusion_char, clean=True)
 
-    for title, domain, tinyurl in results[:4]:
+    for i, info in enumerate(process_urls(bot, trigger, urls)):
+        url, title, domain, tinyurl = info
         message = '[ %s ] - %s' % (title, domain)
         if tinyurl:
             message += ' ( %s )' % tinyurl
         # Guard against responding to other instances of this bot.
         if message != trigger:
             bot.say(message)
+            bot.memory['last_seen_url'][trigger.sender] = url
+
+        # stop at 4th iteration
+        if i == 3:
```

Comment:
Why is this one (auto-title) `enumerate(process_urls([blah]))` and the other (manual command) is `results[:4]`?

I feel like we shouldn't need an `i` counter variable in Python. :/ But I also think both places should use the same pattern, not have one store all URLs in a `results` list and the other stop iteration before processing all URLs.

Reply:
Yeah, but there is a problem:

* we fetch all URLs, no matter how many we display
* we display only 4 of them at most (like it's a common thing)
* in the other function, we try to be smart with an error message if the number of URL does not match the number of result
* one of the function check for malicious links, the other doesn't,
* etc.

I think the differences between these functions are irrelevant and should be restructured into this:

* always display *all* URL, it's not like people will flood with tinyurl on 512 bytes (minus their hostmask)
* display an error message per URL that we fail to get (and not just URL that are handled by a callback)
* apply the same protection on all URLs
* register the "last seen URL" on both .title and auto-url

This way I can fix that `enumerate` problem, and never rely on `list(generator-function)` again. :D